### DISCUSSION TITLE
Theremin design

### DISCUSSION THREAD

**Discussion by juanito76** [marcel-licence/esp32_basic_synth#83] [Ideas]
Hello Marcel, I would like to make an very lowcost educative Theremin for schools. The idea es to have 2 analog sensors, one for changing modulator and other for changing pitch of a selected note. I was playing with your code but I cannot get a "good" comprehension about how it is working and the sound looks terrible. For sure I am using an I2S DAC. 

For this project, no need to have midi input... just select by firmware a note, selecting wafeforms and filters and change mod/pitch or even change filter to reproduce a "like" theremin sound.

ahahaha... my first goal is to have as instrument intro the Star Trek intro, of course!!!

What things must be search on the code to start working?

Very good job you make here!
Best regards,
juanito76

**Comment by marcel-licence** [comment]
Hello,

oh that sounds like a nice idea. I guess you mean the old star trek with Kirk and Spock. 
I think the easiest part might be to control the master volume by an analogue value. I made the adc module in past but you can also make your own implementation. 

The pitch could be a bit tricky. You would need a NoteOn on startup and you will find in the NoteOn function that the pitch is calculated by the note number. It should be possible to update the pitch with a float value in a certain range.

I think I could provide maybe a prototype function soon.

Best regards,
Marcel

**Comment by marcel-licence** [comment]
I tried starting to apply two controls (volume and pitch).

Due to a little bug in my implementation it sounds like I've created a new style of music
https://user-images.githubusercontent.com/65245426/174475276-f8b85e01-7f28-4400-ab94-8f0bc59b7d47.mov

The accuracy of 1024 steps is also a bit rough for pitch control. I tried to use an infrared based IR distance sensor for pitch control. I can keep you updated

**Reply by juanito76** [reply]
Hello Marcel! I will test some day of the following week. By now, I share the code where get an value from an LDR and pass it (raw...) to Synth_PitchBend function:

void Core0TaskLoop()
{
    /*
     * put your loop stuff for core0 here
     */
  #ifdef ADC_TO_MIDI_ENABLED              
    //AdcMul_Process();
  #endif
   //readSimplePots();
   #ifdef DISPLAY_1306
   // displayRefresh();
   #endif
  #ifdef THEREMIN
   //processButtons();
   if(millis()-t_ldr>LDR_MSECS) {
      t_ldr=millis();
      light_value=analogRead(34);
      float val = (float)light_value;
      Synth_PitchBend(0,1/(val/4096));
      
   }
   
  #else
    //probando cosas...
    Synth_NoteOn(0, 80,1.0f);
  #endif
}


The output sound of "this" is the one of the attached video. (sorry about quality, I have to reduce size with ffmpeg and it seems to drop many frames at the beginning, audio is original one)

Thanks a lot!



https://user-images.githubusercontent.com/106962224/174478923-a306d25a-dc84-49fa-a63a-b983bae051dd.mp4

**Reply by marcel-licence** [reply]
Thank you for sharing. It seems you do not have many noise on the ADC. Using the ESP32 Audio Kit shows a lot of noise on the ADC channel. I try to get rid of it by filtering the value. I used the Core0Loop to get the value. And update the PitchBend in the loop() function using a filtered value. 

I put in a value between -24 up to 48 for pitch bend to have a very big range. 

For volume I used the pressure sensor. Maybe I could find something different. You can add the volume control with the code like this:

```
    for (uint32_t n = 0; n < SAMPLE_BUFFER_SIZE; n++)
    {
        pressureValueFilt = pressureValueFilt * 0.99f + pressureValue * 0.01f;
        fl_sample[n] *= pressureValueFilt;
        fr_sample[n] *= pressureValueFilt;
    }

    /*
     * Output the audio
     */
    Audio_Output(fl_sample, fr_sample);
```

The filter might be not required in your build in case the ADC does not collect to much noise otherwise you will get a very rough sound.
